### PR TITLE
executor: streaming stats api

### PR DESCRIPTION
**Pull request by nickethier**
This PR updates the executor and driver APIs to support streaming stats to the client.

### CODE REVIEW

File: drivers/shared/executor/client.go [hunk 1]
```diff
@@ -100,19 +101,42 @@ func (c *grpcExecutorClient) Version() (*ExecutorVersion, error) {
...
+	go c.handleStats(ctx, stream, ch)
+	return ch, nil
+}
 
+func (c *grpcExecutorClient) handleStats(ctx context.Context, stream proto.Executor_StatsClient, ch chan *cstructs.TaskResourceUsage) {
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			if err != io.EOF {
+				// XXX handle error
```
Comment (schmichael) [inline drivers/shared/executor/client.go]: Log and return?

File: drivers/shared/executor/client.go [hunk 2]
```diff
@@ -100,19 +101,42 @@ func (c *grpcExecutorClient) Version() (*ExecutorVersion, error) {
...
+				continue
+			}
+
+			// End stream
+			return
+		}
+
+		stats, err := drivers.TaskStatsFromProto(resp.Stats)
+		if err != nil {
+			// XXX handle error
```
Comment (schmichael) [inline drivers/shared/executor/client.go]: Log and return?

File: drivers/shared/executor/client.go [hunk 3]
```diff
@@ -100,19 +101,42 @@ func (c *grpcExecutorClient) Version() (*ExecutorVersion, error) {
...
-		return nil, err
-	}
-	return stats, nil
+	ch := make(chan *cstructs.TaskResourceUsage)
+	go c.handleStats(ctx, stream, ch)
+	return ch, nil
+}
 
+func (c *grpcExecutorClient) handleStats(ctx context.Context, stream proto.Executor_StatsClient, ch chan *cstructs.TaskResourceUsage) {
+	for {
```
Comment (schmichael) [inline drivers/shared/executor/client.go]: `defer close(ch)`?

File: drivers/shared/executor/client.go [hunk 4]
```diff
@@ -100,19 +101,42 @@ func (c *grpcExecutorClient) Version() (*ExecutorVersion, error) {
...
-	if err != nil {
-		return nil, err
-	}
-	return stats, nil
+	ch := make(chan *cstructs.TaskResourceUsage)
+	go c.handleStats(ctx, stream, ch)
+	return ch, nil
+}
 
+func (c *grpcExecutorClient) handleStats(ctx context.Context, stream proto.Executor_StatsClient, ch chan *cstructs.TaskResourceUsage) {
```
Comment (schmichael) [inline drivers/shared/executor/client.go]:
```suggestion
func (c *grpcExecutorClient) handleStats(ctx context.Context, stream proto.Executor_StatsClient, ch chan<- *cstructs.TaskResourceUsage) {
```

File: drivers/shared/executor/executor.go [hunk 5]
```diff
@@ -507,12 +508,33 @@ func (e *UniversalExecutor) Signal(s os.Signal) error {
...
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-timer.C:
+			timer.Reset(interval)
+		}
+
+		pidStats, _ := e.pidCollector.pidStats()
+		// XXX handle error
```
Comment (schmichael) [inline drivers/shared/executor/executor.go]: Log and return?

File: drivers/shared/executor/executor_linux.go [hunk 6]
```diff
@@ -355,60 +357,84 @@ func (l *LibcontainerExecutor) Version() (*ExecutorVersion, error) {
...
+		}
+
+		lstats, err := l.container.Stats()
+		if err != nil {
+			// XXX handle error
+		}
+
+		pidStats, err := l.pidCollector.pidStats()
+		if err != nil {
+			// XXX handle error
```
Comment (schmichael) [inline drivers/shared/executor/executor_linux.go]: Log and return?

File: drivers/shared/executor/executor_linux.go [hunk 7]
```diff
@@ -355,60 +357,84 @@ func (l *LibcontainerExecutor) Version() (*ExecutorVersion, error) {
...
+			return
 
-	return &taskResUsage, nil
+		case <-timer.C:
+			timer.Reset(interval)
+		}
+
+		lstats, err := l.container.Stats()
+		if err != nil {
+			// XXX handle error
```
Comment (schmichael) [inline drivers/shared/executor/executor_linux.go]: Log and return?

File: drivers/shared/executor/server.go [hunk 8]
```diff
@@ -87,20 +87,46 @@ func (s *grpcExecutorServer) Version(context.Context, *proto.VersionRequest) (*p
...
+		case resp, ok := <-outCh:
+			// chan closed, end stream
+			if !ok {
+				return nil
+			}
+
+			// XXX handle wrapped error
+
+			pbStats, _ := drivers.TaskStatsToProto(resp)
+			// XXX wrap error
```
Comment (schmichael) [inline drivers/shared/executor/server.go]: Can't you return errors here?

File: plugins/drivers/client.go [hunk 9]
```diff
@@ -257,20 +258,48 @@ func (d *driverPluginClient) InspectTask(taskID string) (*TaskStatus, error) {
...
+		}
+
+		stats, err := TaskStatsFromProto(resp.Stats)
+		if err != nil {
+			ch <- &cstructs.TaskResourceUsage{
+				Err: fmt.Errorf("failed to decode stats from RPC: %v", err),
+			}
+		}
+
+		ch <- stats
```
Comment (schmichael) [inline plugins/drivers/client.go]:
```go
select {
case <-ctx.Done():
case ch <- stats:
}
```
Do we need this and likewise for where we return errors?

File: client/allocrunner/taskrunner/stats_hook.go [hunk 10]
```diff
@@ -48,13 +49,14 @@ func (h *statsHook) Poststart(ctx context.Context, req *interfaces.TaskPoststart
...
-	if h.stopCh != nil {
+	if h.cancel != nil {
 		h.logger.Debug("poststart called twice without exiting between")
-		close(h.stopCh)
+		h.cancel()
 	}
 
-	h.stopCh = make(chan struct{})
-	go h.collectResourceUsageStats(req.DriverStats, h.stopCh)
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (schmichael) [inline client/allocrunner/taskrunner/stats_hook.go]: Using the passed-in context means this context will get canceled when the task is killed.

That seems fine -- even optimal probably -- but it's such a subtle change I wanted to call it out in case anybody can think of a reason it shouldn't be this way. :)

File: drivers/docker/stats.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+		statsOpts := docker.StatsOptions{
+			ID:      h.containerID,
+			Context: ctx,
+			Done:    h.doneCh,
+			Stats:   statsCh,
+			Stream:  true,
+		}
+
+		// Stats blocks until an error has occurred, or doneCh has been closed
+		if err := h.client.Stats(statsOpts); err != nil && err != io.ErrClosedPipe {
```
Comment (schmichael) [inline drivers/docker/stats.go]: ErrClosedPipe indicates the Read-side closed, but do we also need to ignore io.EOF for when the Write-side closes?

File: drivers/docker/stats_test.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	stats.MemoryStats.Stats.Cache = 1234
+	stats.MemoryStats.Stats.Swap = 0
+	stats.MemoryStats.MaxUsage = 6651904
+
+	go dockerStatsCollector(dst, src, time.Second)
+
+	select {
+	case src <- stats:
+	case <-time.After(time.Second):
+		require.Fail("sending stats should not block here")
```
Comment (schmichael) [inline drivers/docker/stats_test.go]: TIL there's a require.Fail func! I've still been using t.Fatal in these cases but will stop now. :man_facepalming:

File: drivers/qemu/driver_test.go [hunk 13]
```diff
@@ -355,16 +355,19 @@ func TestQemuDriver_Stats(t *testing.T) {
...
-	// Ask for stats again
-	stats, err = harness.TaskStats(task.ID)
-	require.NoError(err)
-
-	fmt.Printf("CPU:%+v Memory:%+v\n", stats.ResourceUsage.CpuStats, stats.ResourceUsage.MemoryStats)
-	require.NotZero(stats.ResourceUsage.MemoryStats.RSS)
-	require.NoError(harness.DestroyTask(task.ID, true))
+	select {
+	case stats := <-statsCh:
+		fmt.Printf("CPU:%+v Memory:%+v\n", stats.ResourceUsage.CpuStats, stats.ResourceUsage.MemoryStats)
```
Comment (schmichael) [inline drivers/qemu/driver_test.go]: Might as well fix this up while we're here.
```suggestion
		t.Logf("CPU:%+v Memory:%+v\n", stats.ResourceUsage.CpuStats, stats.ResourceUsage.MemoryStats)
```

File: drivers/rkt/driver_test.go [hunk 14]
```diff
@@ -861,12 +861,19 @@ func TestRktDriver_Stats(t *testing.T) {
...
 	require.Nil(err)
 
-	//TODO(preetha) why are these zero
-	fmt.Printf("pid map %v\n", resourceUsage.Pids)
-	fmt.Printf("CPU:%+v Memory:%+v", resourceUsage.ResourceUsage.CpuStats, resourceUsage.ResourceUsage.MemoryStats)
+	select {
+	case ru := <-statsCh:
+		//TODO(preetha) why are these zero
+		fmt.Printf("pid map %v\n", ru.Pids)
+		fmt.Printf("CPU:%+v Memory:%+v", ru.ResourceUsage.CpuStats, ru.ResourceUsage.MemoryStats)
```
Comment (schmichael) [inline drivers/rkt/driver_test.go]: Well...... Let's just save these for another day I guess.

File: drivers/shared/executor/executor_test.go [hunk 15]
```diff
@@ -77,6 +77,8 @@ func TestExecutor_Start_Invalid(pt *testing.T) {
 	pt.Parallel()
 	invalid := "/bin/foobar"
 	for name, factory := range executorFactories {
+		name := name
+		factory := factory
```
Comment (schmichael) [inline drivers/shared/executor/executor_test.go]: I don't think this is necessary because the _subtest_ doesn't call t.Parallel, but either way feel free to leave it.

File: drivers/shared/executor/executor_test.go [hunk 16]
```diff
@@ -167,14 +175,33 @@ func TestExecutor_WaitExitSignal(pt *testing.T) {
...
-				time.Sleep(2 * time.Second)
-				_, err := executor.Stats()
-				require.NoError(err)
-				//require.NotEmpty(ru.Pids)
-				proc, err := os.FindProcess(ps.Pid)
-				require.NoError(err)
-				err = proc.Signal(syscall.SIGKILL)
-				require.NoError(err)
+				// Give process time to start
+				time.Sleep(time.Second)
```
Comment (schmichael) [inline drivers/shared/executor/executor_test.go]: You can probably remove this since WaitForResult retries for up to 5s.

File: drivers/shared/executor/executor_test.go [hunk 17]
```diff
@@ -167,14 +175,33 @@ func TestExecutor_WaitExitSignal(pt *testing.T) {
...
+					if err != nil {
+						return false, err
+					}
+					err = proc.Signal(syscall.SIGKILL)
+					if err != nil {
+						return false, err
+					}
+					return true, nil
+				}, func(err error) {
+					require.NoError(err)
```
Comment (schmichael) [inline drivers/shared/executor/executor_test.go]: Shouldn't Fatal inside a goroutine.

File: drivers/shared/executor/executor_test.go [hunk 18]
```diff
@@ -289,7 +318,7 @@ func TestUniversalExecutor_LookupPath(t *testing.T) {
 // than mounting the underlying host filesystem
 func setupRootfs(t *testing.T, rootfs string) {
 	paths := []string{
-		"/bin/sh",
+		//		"/bin/sh",
```
Comment (schmichael) [inline drivers/shared/executor/executor_test.go]: Why did you remove /bin/sh? Shouldn't it be permanently removed?

File: drivers/shared/executor/server.go [hunk 19]
```diff
@@ -87,20 +87,46 @@ func (s *grpcExecutorServer) Version(context.Context, *proto.VersionRequest) (*p
...
 
-func (s *grpcExecutorServer) Stats(context.Context, *proto.StatsRequest) (*proto.StatsResponse, error) {
-	stats, err := s.impl.Stats()
-	if err != nil {
-		return nil, err
+func (s *grpcExecutorServer) Stats(req *proto.StatsRequest, stream proto.Executor_StatsServer) error {
+	ctx := stream.Context()
+
+	interval := time.Duration(req.Interval)
+	if interval.Nanoseconds() == 0 {
```
Comment (schmichael) [inline drivers/shared/executor/server.go]: The .Nanoseconds() call is unnecessary when comparing against a const, but I'm sure the compiler emits the same code either way.

File: plugins/drivers/server.go [hunk 20]
```diff
@@ -221,22 +222,36 @@ func (b *driverPluginServer) InspectTask(ctx context.Context, req *proto.Inspect
...
+			return st.Err()
+		}
+		return err
 	}
 
-	pb, err := TaskStatsToProto(stats)
-	if err != nil {
-		return nil, fmt.Errorf("failed to encode task stats: %v", err)
-	}
+	for stats := range ch {
```
Comment (schmichael) [inline plugins/drivers/server.go]: The few times ranging over a chan works it's so satisfying.

File: drivers/docker/stats.go (hunk 11 above)
Comment (nickethier) [reply drivers/docker/stats.go]: We actually don't get an EOF. The docker api catches that and returns nil since thats the indicator that the container has stopped.

File: drivers/shared/executor/executor_test.go (hunk 15 above)
Comment (nickethier) [reply drivers/shared/executor/executor_test.go]: Yeah I get really jumpy when I see sub tests in a for loop. There may be some historic trauma debugging a rouge t.Parallel :grimacing:

File: drivers/shared/executor/executor_test.go (hunk 18 above)
Comment (nickethier) [reply drivers/shared/executor/executor_test.go]: Oh drat this change wasn't suppose to make it in.

File: client/allocrunner/taskrunner/stats_hook.go (hunk 10 above)
Comment (nickethier) [reply client/allocrunner/taskrunner/stats_hook.go]: So before stats was signaled to stop when the Exited hook ran, rather than when the task was killed. Hmm then nilling out the cancel is probably over kill then? Can we just pass the context all together without a child cancel and get rid of the Exited hook?

**Comment by nickethier** [comment]
@schmichael ready for another look

File: client/allocrunner/taskrunner/stats_hook.go [hunk 21]
```diff
@@ -22,8 +23,8 @@ type statsHook struct {
 	updater  StatsUpdater
 	interval time.Duration
 
-	// stopCh is closed by Exited or Canceled
-	stopCh chan struct{}
+	// cancel is called by Exited or Canceled
```
Comment (schmichael) [inline client/allocrunner/taskrunner/stats_hook.go]: Ha, I don't even remember what "Canceled" was, but it doesn't appear to exist anymore.
```suggestion
	// cancel is called by Exited
```

File: client/allocrunner/taskrunner/stats_hook.go (hunk 10 above)
Comment (schmichael) [reply client/allocrunner/taskrunner/stats_hook.go]: Ugh, this is a good question, and I believe the answer has changed over time.

Initially we only used contexts for hooks like a request/response: useful for cancelling long running hooks, but *not* for signalling task state changes.

Now we can signal task state changes for hooks with background goroutines through a chan. This is convenient, but it's an emergent property of the API, not an intentional design choice.

I _think_ we ideally want to treat contexts as being request scoped (where "request" == a hook method), so hooks operate like external plugins (as we might want to make them plugins in the future).

If we want to enforce that we should wrap Poststart's context is a subcontext that is cancelled after each call to avoid developers using it instead of Exited.
